main.c

```diff
@@ -1,3 +1,44 @@
+#include <stdint.h>
+
+#include "am1808.h"
+
+// Shared RAM (control input)
+typedef struct shared_ram {
+    // Declare this as a u32 to force more-efficient codegen
+    uint32_t pwms;
+} shared_ram;
+static volatile shared_ram * const SHARED = (volatile shared_ram *)0x80010000;
+// XXX this doesn't seem to work (compiler won't generate code to use it)
+// #pragma ctable_entry 30 0x80010000
+
+// LED definitions (all on GP6)
+#define LED0    (1 << 13)
+#define LED1    (1 << 7)
+#define LED2    (1 << 14)
+#define LED3    (1 << 12)
+
+static inline void update_pwm(uint8_t val, uint8_t time_now, uint32_t gpio_bit) {
+    if (time_now < val) {
+        GPIO->set_data67 = gpio_bit;
```

Comment:
If we are playing the game of save every byte we can, what is smaller, using the R30 to access the GPIOs or this?

Assuming we are using PRU1, we should be able to use 

| | |
|-|-|
| LED0 | R30[12] |
| LED1 | R30[10] |
| LED2 | R30[13] |
| LED3 | R30[11] |

If we set the pinmux for it.

Reply:
You must have been looking at PRU0. :smile: